### DISCUSSION TITLE
Many new features are launching in SDWEBUI extension and standalong gradio demo

### DISCUSSION THREAD

**Discussion by vinthony** [OpenTalker/SadTalker#386] [Announcements]
Hi, everyone! 

Thanks for your patience with the bugs and long-time updates in SadTalker.

We are releasing some new features in SadTalker for the WEBUI.

You can try to install the dev version for these new features:
```
git clone -b dev https://github.com/OpenTalker/SadTalker
```

OR in SDWEBUI:
```
cd extensions/SadTalker && git checkout dev
```

It will be merged into the main code after careful testing.

#### 1. The IDLE mode as discussed in https://github.com/OpenTalker/SadTalker/issues/280#issuecomment-1559038384.

Now, you can simply generate the video with a simple image from SD or by yourself with the length of the video as input. 
Where the head pose will be generated automatically as well as the eye blinking.
![image](https://github.com/OpenTalker/SadTalker/assets/4397546/0d5bb3e0-5d5d-44cf-82a4-ee5b3e9620cd)

#### 2. Ref Video mode.

It also allows a new video as reference pose and eye blink and **FULLY MOTION**! 

- As for the reference pose and eye blink, the generated video still needs the audio from IDLE/AUDIO FILE/TTS.
- As for the full motion mode(all), the generated video will use the audio from the input video and generate the same motion as the original video.

![image](https://github.com/OpenTalker/SadTalker/assets/4397546/7f42d908-e098-40de-90c5-78ed13edeef6)

 
#### 3. Controllable Blinking

We set a new use blink checkbox in settings to control the eye blink explicitly. The generated video in FULL mode may work fine without an eye blink.

![image](https://github.com/OpenTalker/SadTalker/assets/4397546/8fe056d3-f4d0-4db7-8fd2-6267309b296e)

Have fun and let me know if you have a better idea or problem with these updates.

**Comment by james-s-tayler** [comment]
I pulled the latest version of the extension but I don't see these features? There's no UI for enabling the blink checkbox, and I can't see anything about reference video either? Am I missing something?

**Reply by vinthony** [reply]
to avoid bugs i have make a dev branch to test these new features. you may need to checkout the dev branch as below:

<img width="911" alt="image" src="https://github.com/OpenTalker/SadTalker/assets/4397546/d3a0fac5-48be-42b1-80dd-948a20996341">

**Comment by james-s-tayler** [comment]
I already looked at the `dev` branch and saw it only contained a single commit with a single one line change, so I didn't bother checking it out because it only added a print statement. Have the changes actually been pushed to that branch?

**Reply by vinthony** [reply]
Sure, the code are here in dev branch:
https://github.com/OpenTalker/SadTalker/blob/4c38d1f5950a5fc2616f26d7e92e750f51c13ac2/app.py#L45-L100

**Comment by JT-Apps** [comment]
> I pulled the latest version of the extension but I don't see these features? There's no UI for enabling the blink checkbox, and I can't see anything about reference video either? Am I missing something?



> I pulled the latest version of the extension but I don't see these features? There's no UI for enabling the blink checkbox, and I can't see anything about reference video either? Am I missing something?

I'm having the exact same thing, stoked for these features, did _cd extensions/SadTalker && git checkout dev_ and all but no new interface, after restarts, relaunching, trashing, and reinstalling. Not sure what to do @vinthony...? Thanks for your work on this.

![Screenshot 2023-06-19 at 10 21 32 PM](https://github.com/OpenTalker/SadTalker/assets/106498013/f5f2cf24-cd56-4d31-839f-cf3b6c4b6817)
<img width="1997" alt="Screenshot 2023-06-19 at 10 21 27 PM" src="https://github.com/OpenTalker/SadTalker/assets/106498013/28b5954a-3aba-499c-8ac6-3e9276c62f41">

**Reply by vinthony** [reply]
<img width="461" alt="image" src="https://github.com/OpenTalker/SadTalker/assets/4397546/c39432a3-f716-43f6-a785-f659ba95c901">

Have you see this information when you switch the branch?

**Comment by jjmlovesgit** [comment]
Nice --- and of course I had just done one myself to let me add the pose videos and pose style :)  Guess I will not be needing this Afterall!  

I also was messing with the Yaw, Pitch, and Roll settings... I eventually landed on something I like for my use case where I use a virtual human to walk my code on my free Youtube.  This guy is a great assistant as he never stumbles or mumbles during long code reviews... it definitely is good result off a single still image.  An incredible job by the team involved here.

Here is my baseline I am trying to emulate (Note the watermark as this is paid generation):
https://github.com/OpenTalker/SadTalker/assets/47751509/4301911e-dc1d-4a1c-93a8-f94169af3661

Here is what I was able to generate (This is free SadTalker done locally):
https://github.com/OpenTalker/SadTalker/assets/47751509/8bbce1a9-44d3-433e-a9d3-4004161e4b7f


![image](https://github.com/OpenTalker/SadTalker/assets/47751509/57dcb76d-fe7c-4d9f-b55f-0035299f692c)

**Reply by ldavis9000aws** [reply]
If you're using yaw, pitch, and roll, it sounds like that is for the CLI version while most folks utilize the WebUi that doesn't have those options.

**Comment by G-force78** [comment]
Cool, good to see things get integrated and driving videos possible too like in video-retalking
I did get this error on first run
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/gradio/routes.py", line 437, in run_predict
    output = await app.get_blocks().process_api(
  File "/usr/local/lib/python3.10/dist-packages/gradio/blocks.py", line 1352, in process_api
    result = await self.call_function(
  File "/usr/local/lib/python3.10/dist-packages/gradio/blocks.py", line 1077, in call_function
    prediction = await anyio.to_thread.run_sync(
  File "/usr/local/lib/python3.10/dist-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/usr/local/lib/python3.10/dist-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/usr/local/lib/python3.10/dist-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "/content/SadTalker/src/gradio_demo.py", line 107, in test
    ref_video_coeff_path, _, _ =  self.preprocess_model.generate(ref_video, ref_video_frame_dir, preprocess, source_image_flag=False)
  File "/content/SadTalker/src/utils/preprocess.py", line 124, in generate
    lm = self.propress.predictor.extract_keypoint(frames_pil, landmarks_path)
  File "/content/SadTalker/src/face3d/extract_kp_videos_safe.py", line 39, in extract_keypoint
    current_kp = self.extract_keypoint(image)
  File "/content/SadTalker/src/face3d/extract_kp_videos_safe.py", line 57, in extract_keypoint
    bboxes = bboxes[0]
IndexError: index 0 is out of bounds for axis 0 with size 0
```

Ok so i changed it from pose and blink to full , reduced pose scale and it started generating but then got this error

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/gradio/routes.py", line 437, in run_predict
    output = await app.get_blocks().process_api(
  File "/usr/local/lib/python3.10/dist-packages/gradio/blocks.py", line 1352, in process_api
    result = await self.call_function(
  File "/usr/local/lib/python3.10/dist-packages/gradio/blocks.py", line 1077, in call_function
    prediction = await anyio.to_thread.run_sync(
  File "/usr/local/lib/python3.10/dist-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/usr/local/lib/python3.10/dist-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/usr/local/lib/python3.10/dist-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "/content/SadTalker/src/gradio_demo.py", line 139, in test
    return_path = self.animate_from_coeff.generate(data, save_dir,  pic_path, crop_info, enhancer='gfpgan' if use_enhancer else None, preprocess=preprocess, img_size=size)
  File "/content/SadTalker/src/facerender/animate.py", line 215, in generate
    sound = AudioSegment.from_file(audio_path)
  File "/usr/local/lib/python3.10/dist-packages/pydub/audio_segment.py", line 651, in from_file
    file, close_file = _fd_or_path_or_tempfile(file, 'rb', tempfile=False)
  File "/usr/local/lib/python3.10/dist-packages/pydub/utils.py", line 60, in _fd_or_path_or_tempfile
    fd = open(fd, mode=mode)
FileNotFoundError: [Errno 2] No such file or directory: './results/5ef6ef02-53f6-4e75-9263-4ca6d8c07299/Patrick Stewart is So Proud of Star Trek Picard Season 3 After Initial Worries - TheWrap Magazine_Clip 2.mp4.wav'
```
Here are my settings used

![webuisadsettings](https://github.com/OpenTalker/SadTalker/assets/114336644/a1276867-f828-4d50-ade9-93328f5c5f68)

I have now been able to generate by using these settings. I did also add some code to enable gradio to setup tmp files, no idea if that did anything
```
def upload_files(input_file):
    # Create a temporary folder
    tmp_folder = tempfile.mkdtemp()

    # Save the uploaded file to the temporary folder
    input_file.save(os.path.join(tmp_folder, input_file.name))

    # Process the uploaded file
    # ...

    # Return the result
    return result
```
Settings used for successful result
![webuisadsettings1](https://github.com/OpenTalker/SadTalker/assets/114336644/4c6d72f5-7a5d-4c33-9f67-b37bfb8954b5)

**Reply by arezaei2002** [reply]
can i see your hole code ??

**Comment by G-force78** [comment]
These are great examples how did you do these? https://user-images.githubusercontent.com/4397546/224310754-665eb2dd-aadc-47dc-b1f9-2029a937b20a.mp4 What settings correspond to in the WebUI?

Also what is the setting extfull?

**Reply by qusonny** [reply]
Is the video in the link above made with Sadtalker? Why is it that when I make the video, the head and body always separate? Unless still mode is selected, but still mode head movement is too little for what I want.

**Comment by barbarabax** [comment]
Love it! Thank you

**Comment by G-force78** [comment]
> Is the video in the link above made with Sadtalker? Why is it that when I make the video, the head and body always separate? Unless still mode is selected, but still mode head movement is too little for what I want.

Certainly not from scratch it looks like its just been lip synced using something like wav2lip

**Comment by bolocuixue** [comment]
为什么我生成的人物口型没有牙齿

**Comment by slavakurilyak** [comment]
When can we expect this `dev` code to be merged into `main`?

**Comment by Nate82** [comment]
For anyone else still getting the old UI when you have the latest version installed:

`cd extensions/SadTalker && git checkout dev`

The above line did not work for me, the old UI was still showing.

So I deleted every SadTalker folder/file and ran the following inside \stable-diffusion-webui\extensions

`git clone -b dev https://github.com/OpenTalker/SadTalker`

That downloaded the dev branch directly and created a SadTalker folder inside \stable-diffusion-webui\extensions

Then I created a **checkpoints** folder inside the SadTalker folder and downloaded the 4 files from the latest release on https://github.com/OpenTalker/SadTalker/releases

Now the new updated UI is showing in Automatic1111

**Comment by keith-maltman** [comment]
Hi there,

I'm sorry if this is in the wrong place. I'm VERY new to this. After 3 full days of trying to program to install and run in Ubuntu (Every single set of instructions on the internet AND YOUR OWN has AT LEAST one mistake. (your .sh script does not work in ubuntu BTW), and there is a path error in the manual instructions, (that still did not work once I fixed it?) Anyway I finally managed to work it out the hard way, and get it to install. Only to encounter this message after it attempted to run the first time. New as I am, this time I suspect it's not the install procedure, (that I painfully worked out myself). This time I suspect it something in the app. 

stderr: ERROR: Exception:
Traceback (most recent call last):
  File "/home/lucifer/anaconda3/envs/sadtalker/lib/python3.8/site-packages/pip/_internal/cli/base_command.py", line 180, in exc_logging_wrapper
    status = run_func(*args)
  File "/home/lucifer/anaconda3/envs/sadtalker/lib/python3.8/site-packages/pip/_internal/cli/req_command.py", line 245, in wrapper
    return func(self, options, args)
  File "/home/lucifer/anaconda3/envs/sadtalker/lib/python3.8/site-packages/pip/_internal/commands/install.py", line 377, in run
    requirement_set = resolver.resolve(
  File "/home/lucifer/anaconda3/envs/sadtalker/lib/python3.8/site-packages/pip/_internal/resolution/resolvelib/resolver.py", line 95, in resolve
    result = self._result = resolver.resolve(
  File "/home/lucifer/anaconda3/envs/sadtalker/lib/python3.8/site-packages/pip/_vendor/resolvelib/resolvers.py", line 546, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File "/home/lucifer/anaconda3/envs/sadtalker/lib/python3.8/site-packages/pip/_vendor/resolvelib/resolvers.py", line 457, in resolve
    raise ResolutionTooDeep(max_rounds)
pip._vendor.resolvelib.resolvers.ResolutionTooDeep: 200000

Anyone know what this error is and how to fix it, I suspect it to do with TTS, but I have only been using Linux and python for a couple of weeks.

Keith

Sorry again if this is in the wrong place

**Comment by saurytang** [comment]
启动webuui的时候发生异常，你那个gradio用的那个版本呀？
发生异常: AttributeError
'Row' object has no attribute 'style'
  File "D:\datahuman\SadTalker\app_sadtalker.py", line 35, in sadtalker_demo
    with gr.Row().style(equal_height=False):
  File "D:\datahuman\SadTalker\app_sadtalker.py", line 113, in <module>
    demo = sadtalker_demo()
AttributeError: 'Row' object has no attribute 'style'

**Comment by 1joshw** [comment]
can anyone tell me how to fix
![ss](https://github.com/OpenTalker/SadTalker/assets/142825686/1992ac77-e3a2-43ed-8059-9a60571fde84)

**Comment by saurytang** [comment]
pip install gradio==3.41.2




saury
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "OpenTalker/SadTalker"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年5月12日(星期天) 凌晨1:02
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [OpenTalker/SadTalker] Many new features are launching in SDWEBUI extension and standalong gradio demo (Discussion #386)





 
can anyone tell me how to fix
 ss.png (view on web)

**Reply by 1joshw** [reply]
thanks, but now.
![ss](https://github.com/OpenTalker/SadTalker/assets/142825686/966a7e54-2bef-42a6-8b97-652e23ced09a)

**Reply by Rubenacp** [reply]
I think the instruction pip install gradio==3.41.2 is only for the gradio UI, no for automatic1111 or forge.

**Comment by saurytang** [comment]
saury
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "OpenTalker/SadTalker"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年5月12日(星期天) 晚上11:39
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [OpenTalker/SadTalker] Many new features are launching in SDWEBUI extension and standalong gradio demo (Discussion #386)





 
thanks, but now.
 ss.png (view on web)

**Comment by johndpope** [comment]
hey - 
a new paper dropped - IMTalker - i got it working with 58 fps - 
https://github.com/johndpope/IMTalker

im tinking of upgrading the demo - im wondering what the best ui for presenting 
- i love the idle logic here -

**Reply by jjmlovesgit** [reply]
@johndpope That is mind blowing how fast it is... I thank you for sharing!

**Reply by jjmlovesgit** [reply]
@johndpope Routing an LLM response to tts and on to a FastAPI endpoint with IMTalker to get a near real-time local chatbot in the browser.  Local uncensored LLM etc. where you pick the Avatar image and set the prompt.

Many challenges trying to get a streaming type of experience when just using the browser...we do a sentence by sentence parse into a queue for playback where we start the first sentence .mp4 playing while we work to stay ahead by generating all the rest of the sentences .mp4 as the first one plays... the stitching of the video one after another leads to jumps... the flow requires tuning the prompt to deliver the right length of sentences back where the wait on the first one is not so long you get annoyed.

A IMTalker is a very cool project...
 
https://github.com/user-attachments/assets/80f85e20-7579-4c1e-a4c6-c354136360c9

**Reply by johndpope** [reply]
because under the hood - that technology is using IMF - ( which i successfully recreated - https://github.com/johndpope/IMF )
https://openaccess.thecvf.com/content/CVPR2024/papers/Gao_Implicit_Motion_Function_CVPR_2024_paper.pdf
there is opportunity to actually stream the tokens (32 bytes) across the wire per frame - you give the reference image + and then stream tokens down the wire - it's super efficient - 
<img width="671" height="325" alt="ok" src="https://github.com/user-attachments/assets/5ed959fe-ab46-49fe-8e53-c78b69ce6abb" />

i did have a go with doing onnx conversion (that was a painful experience at the time porting pytorch -> onnx->  tensorflow js) 

i upgrade the docs in imtalker if anyone is interested - https://github.com/johndpope/IMTalker/tree/feat/imf-documentation

**Comment by jjmlovesgit** [comment]
Desktop Agent -- private offline local Avatar with IMTalker with Driver video for damping the head movements


https://github.com/user-attachments/assets/c8932d6d-2401-4bd2-89f8-51cd76f38786

**Reply by johndpope** [reply]
looking at the rendering pipeline and investigating hotswapping this out with .... (checks notes)  triangular splatting + threejs - https://threejs.covershot.app/ 

https://github.com/trianglesplatting2/triangle-splatting2 - its 2400 fps
